### PR TITLE
Add LKJCholesky distribution

### PR DESCRIPTION
**Pull request by neerajprad**
As a follow up to #48041, this adds the `LKJCholesky` distribution that samples the Cholesky factor of positive definite correlation matrices.

This also relaxes the check on `tril_matrix_to_vec` so that it works for 2x2 matrices with `diag=-2`.

cc. @fehiepsi

### CODE REVIEW

File: torch/distributions/lkj_cholesky.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    def sample(self, sample_shape=torch.Size()):
+        # This uses the Onion method, but tere are a few differences from [1] Sec. 3.2:
+        # - This vectorizes the for loop and also works for heterogeneous eta.
+        # - Same algorithm generalizes to n=1.
+        # - The procedure is simplified since we are sampling the cholesky factor of
+        #   the correlation matrix instead of the correlation matrix itself. As such,
+        #   we only need to generate `w`.
+        y = self._beta.sample(sample_shape).unsqueeze(-1)
+        u_normal = torch.randn(self._extended_shape(sample_shape),
```
Comment (neerajprad) [inline torch/distributions/lkj_cholesky.py:80]: @fehiepsi: I found that sampling a full matrix was faster (instead of sampling half and converting it into a lower tril matrix) in pytorch even if we reject half the samples.

Comment (neerajprad) [reply torch/distributions/lkj_cholesky.py:80]: Could you help review the PR? This follows your implementation in NumPyro, but I am not able to add you as a reviewer.

Comment (fehiepsi) [reply torch/distributions/lkj_cholesky.py:89]: Yup, I'll review it later today. :)

File: torch/distributions/lkj_cholesky.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+import math
+
+import torch
+from torch.distributions import constraints, Beta
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all
+
+
+class LKJCholesky(Distribution):
```
Comment (vishwakftw) [inline torch/distributions/lkj_cholesky.py:18]: The name seems weird to me, but maybe it's ok since we already have `LowRankMultivariateNormal`.

File: torch/distributions/lkj_cholesky.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        if dim < 2:
+            raise ValueError(f'Expected dim to be an integer greater than or equal to 2. Found dim={dim}.')
+        self.dim = dim
+        self.concentration, = broadcast_all(concentration)
+        batch_shape = self.concentration.size()
+        event_shape = torch.Size((dim, dim))
+        # This is used to draw vectorized samples from the beta distribution in Sec. 3.2 of [1].
+        marginal_conc = self.concentration + 0.5 * (self.dim - 2)
+        offset = torch.arange(self.dim - 1, dtype=self.concentration.dtype, device=self.concentration.device)
+        offset = torch.cat([offset.new_zeros((1,)), offset])
```
Comment (vishwakftw) [inline torch/distributions/lkj_cholesky.py:65]: Not sure if this is a recommended factory usage. Perhaps `torch.zeros(1, dtype=offset.dtype, device=offset.device)`?

File: test/distributions/test_distributions.py [hunk 4]
```diff
@@ -2279,10 +2295,10 @@ def test_kumaraswamy_shape(self):
...
     # Hence these tests are explicit
     def test_kumaraswamy_mean_variance(self):
-        c1_1 = torch.tensor(torch.randn(2, 3).abs(), requires_grad=True)
-        c0_1 = torch.tensor(torch.randn(2, 3).abs(), requires_grad=True)
-        c1_2 = torch.tensor(torch.randn(4).abs(), requires_grad=True)
-        c0_2 = torch.tensor(torch.randn(4).abs(), requires_grad=True)
+        c1_1 = torch.randn(2, 3).abs().requires_grad_()
+        c0_1 = torch.randn(2, 3).abs().requires_grad_()
+        c1_2 = torch.randn(4).abs().requires_grad_()
+        c0_2 = torch.randn(4).abs().requires_grad_()
```
Comment (vishwakftw) [inline test/distributions/test_distributions.py:2301]: Yikes, sorry about this.

File: test/distributions/test_distributions.py [hunk 5]
```diff
@@ -2265,10 +2281,10 @@ def test_gumbel_sample(self):
...
 
     def test_kumaraswamy_shape(self):
-        concentration1 = torch.tensor(torch.randn(2, 3).abs(), requires_grad=True)
-        concentration0 = torch.tensor(torch.randn(2, 3).abs(), requires_grad=True)
-        concentration1_1d = torch.tensor(torch.randn(1).abs(), requires_grad=True)
-        concentration0_1d = torch.tensor(torch.randn(1).abs(), requires_grad=True)
+        concentration1 = torch.randn(2, 3).abs().requires_grad_()
+        concentration0 = torch.randn(2, 3).abs().requires_grad_()
+        concentration1_1d = torch.randn(1).abs().requires_grad_()
+        concentration0_1d = torch.randn(1).abs().requires_grad_()
```
Comment (vishwakftw) [inline test/distributions/test_distributions.py:2287]: And this.

**Review by vishwakftw — COMMENTED** [review]
I haven't checked the math, mostly minor comments.

File: torch/distributions/lkj_cholesky.py (hunk 2 above)
Comment (neerajprad) [reply torch/distributions/lkj_cholesky.py:18]: The `LKJ` or `LKJCorr` distribution in the paper samples from the space of positive definite correlation matrices, but this one samples the Cholesky factor of correlation matrices. The name is to help clarify this. TFP calls this `CholeskLKJ` and Stan calls this `lkj_corr_cholesky`. I'm happy to change it to something else if you have any suggestions.

File: test/distributions/test_distributions.py (hunk 4 above)
Comment (neerajprad) [reply test/distributions/test_distributions.py:2301]: Np, I just noticed that due to some warnings.

File: torch/distributions/lkj_cholesky.py (hunk 3 above)
Comment (neerajprad) [reply torch/distributions/lkj_cholesky.py:65]: Hmm..we still have these in the official docs - https://pytorch.org/docs/stable/tensors.html. Is there a plan for deprecating these constructors?

**Comment by neerajprad** [comment]
> I haven't checked the math, mostly minor comments.

Thanks for reviewing, @vishwakftw! @fehiepsi  has agreed to review the math since this is basically a port of his implementation in numpyro.

File: torch/distributions/lkj_cholesky.py (hunk 2 above)
Comment (vishwakftw) [reply torch/distributions/lkj_cholesky.py:18]: Thanks for the explanation, in lieu of this we could retain the name.

File: torch/distributions/lkj_cholesky.py (hunk 3 above)
Comment (vishwakftw) [reply torch/distributions/lkj_cholesky.py:65]: I remember coming across this sometime back, but I might be misremembering. I think we can change when it's deprecated (if it is) Sorry!

**Review by fehiepsi — APPROVED** [review]
LGTM! It is a bit tricky to review the math of this distribution even that we have a version to compare. :)) The test looks good to me.

**Comment by neerajprad** [comment]
Thanks for reviewing, @fehiepsi, @vishwakftw!

File: torch/distributions/lkj_cholesky.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        u_normal = torch.randn(self._extended_shape(sample_shape),
+                               dtype=y.dtype,
+                               device=y.device).tril(-1)
+        u_hypersphere = u_normal / u_normal.norm(dim=-1, keepdim=True)
+        # Replace NaNs in first row
+        u_hypersphere[..., 0, :].fill_(0.)
+        w = torch.sqrt(y) * u_hypersphere
+        # Fill diagonal elements; clamp for numerical stability
+        eps = torch.finfo(w.dtype).tiny
+        diag_elems = torch.clamp(1 - torch.sum(w**2, dim=-1), min=eps).sqrt()
```
Comment (neerajprad) [inline torch/distributions/lkj_cholesky.py:98]: @fehiepsi - I changed min to `torch.finfo(w.dtype).tiny` because otherwise I was seeing samples whose log density would evaluate to inf (specially when concentration param is low).

**Comment by neerajprad** [comment]
@fehiepsi - I was running NUTS using this and made two small changes. Rest everything should be the same.
 - relaxed the check on `vec_to_tril_matrix` so that diag can be [-n, n) instead of (-n, n). This is needed so that we can use LKJ with dim=2. 
 - added a small fudge factor to the diag elements so that we don't return samples whose log density evaluates to inf.
